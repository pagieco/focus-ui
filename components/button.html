<!DOCTYPE html>
<html lang="en-US">
  <head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width,initial-scale=1">
    <title>Button | Focus UI Docs</title>
    <meta name="description" content="">
    
    
    <link rel="preload" href="/focus-ui/assets/css/0.styles.7c918269.css" as="style"><link rel="preload" href="/focus-ui/assets/js/app.ec7e616f.js" as="script"><link rel="preload" href="/focus-ui/assets/js/2.f856ce09.js" as="script"><link rel="preload" href="/focus-ui/assets/js/10.89c71ea6.js" as="script"><link rel="prefetch" href="/focus-ui/assets/js/11.51ec99ee.js"><link rel="prefetch" href="/focus-ui/assets/js/12.696b690b.js"><link rel="prefetch" href="/focus-ui/assets/js/13.e6a8f5e3.js"><link rel="prefetch" href="/focus-ui/assets/js/14.99473aba.js"><link rel="prefetch" href="/focus-ui/assets/js/15.33f3f7ef.js"><link rel="prefetch" href="/focus-ui/assets/js/16.e175e0d8.js"><link rel="prefetch" href="/focus-ui/assets/js/17.92fb1298.js"><link rel="prefetch" href="/focus-ui/assets/js/18.0edd0496.js"><link rel="prefetch" href="/focus-ui/assets/js/19.2ed41cc3.js"><link rel="prefetch" href="/focus-ui/assets/js/20.bbc97549.js"><link rel="prefetch" href="/focus-ui/assets/js/21.e60a114a.js"><link rel="prefetch" href="/focus-ui/assets/js/22.d3bbd8fc.js"><link rel="prefetch" href="/focus-ui/assets/js/23.ce398370.js"><link rel="prefetch" href="/focus-ui/assets/js/24.ed77a66d.js"><link rel="prefetch" href="/focus-ui/assets/js/25.4fdc0e63.js"><link rel="prefetch" href="/focus-ui/assets/js/26.08cb5c31.js"><link rel="prefetch" href="/focus-ui/assets/js/27.5414d0b8.js"><link rel="prefetch" href="/focus-ui/assets/js/28.eb3beb40.js"><link rel="prefetch" href="/focus-ui/assets/js/29.914015b8.js"><link rel="prefetch" href="/focus-ui/assets/js/3.40754a46.js"><link rel="prefetch" href="/focus-ui/assets/js/30.fb1c4726.js"><link rel="prefetch" href="/focus-ui/assets/js/31.978a5344.js"><link rel="prefetch" href="/focus-ui/assets/js/32.6c9df44b.js"><link rel="prefetch" href="/focus-ui/assets/js/33.4f3ddca0.js"><link rel="prefetch" href="/focus-ui/assets/js/34.e1d7f155.js"><link rel="prefetch" href="/focus-ui/assets/js/35.db326522.js"><link rel="prefetch" href="/focus-ui/assets/js/36.4470a8fe.js"><link rel="prefetch" href="/focus-ui/assets/js/37.e3d7fa2b.js"><link rel="prefetch" href="/focus-ui/assets/js/38.fc133bf8.js"><link rel="prefetch" href="/focus-ui/assets/js/39.6ec60546.js"><link rel="prefetch" href="/focus-ui/assets/js/4.11d613bd.js"><link rel="prefetch" href="/focus-ui/assets/js/40.7967dec1.js"><link rel="prefetch" href="/focus-ui/assets/js/41.e4d9d7b4.js"><link rel="prefetch" href="/focus-ui/assets/js/42.bcb427ce.js"><link rel="prefetch" href="/focus-ui/assets/js/43.9c91ae94.js"><link rel="prefetch" href="/focus-ui/assets/js/44.50e43115.js"><link rel="prefetch" href="/focus-ui/assets/js/45.533789bc.js"><link rel="prefetch" href="/focus-ui/assets/js/46.116d38c9.js"><link rel="prefetch" href="/focus-ui/assets/js/5.2718a8ea.js"><link rel="prefetch" href="/focus-ui/assets/js/6.bb50d40e.js"><link rel="prefetch" href="/focus-ui/assets/js/7.dee98ecb.js"><link rel="prefetch" href="/focus-ui/assets/js/8.374f9f82.js"><link rel="prefetch" href="/focus-ui/assets/js/9.4b4b3660.js">
    <link rel="stylesheet" href="/focus-ui/assets/css/0.styles.7c918269.css">
  </head>
  <body>
    <div id="app" data-server-rendered="true"><div class="theme-container"><header class="navbar"><div class="sidebar-button"><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" role="img" viewBox="0 0 448 512" class="icon"><path fill="currentColor" d="M436 124H12c-6.627 0-12-5.373-12-12V80c0-6.627 5.373-12 12-12h424c6.627 0 12 5.373 12 12v32c0 6.627-5.373 12-12 12zm0 160H12c-6.627 0-12-5.373-12-12v-32c0-6.627 5.373-12 12-12h424c6.627 0 12 5.373 12 12v32c0 6.627-5.373 12-12 12zm0 160H12c-6.627 0-12-5.373-12-12v-32c0-6.627 5.373-12 12-12h424c6.627 0 12 5.373 12 12v32c0 6.627-5.373 12-12 12z"></path></svg></div> <a href="/focus-ui/" class="home-link router-link-active"><!----> <span class="site-name">Focus UI Docs</span></a> <div class="links"><div class="search-box"><input aria-label="Search" autocomplete="off" spellcheck="false" value=""> <!----></div> <!----></div></header> <div class="sidebar-mask"></div> <aside class="sidebar"><!---->  <ul class="sidebar-links"><li><a href="/focus-ui/" class="sidebar-link">Introduction</a></li><li><section class="sidebar-group collapsable depth-0"><p class="sidebar-heading open"><span>Components</span> <span class="arrow down"></span></p> <ul class="sidebar-links sidebar-group-items"><li><a href="/focus-ui/components/avatar.html" class="sidebar-link">Avatar</a></li><li><a href="/focus-ui/components/badge.html" class="sidebar-link">Badge</a></li><li><a href="/focus-ui/components/banner.html" class="sidebar-link">Banner</a></li><li><a href="/focus-ui/components/button.html" class="active sidebar-link">Button</a><ul class="sidebar-sub-headers"><li class="sidebar-sub-header"><a href="/focus-ui/components/button.html#best-practices" class="sidebar-link">Best practices</a></li><li class="sidebar-sub-header"><a href="/focus-ui/components/button.html#examples" class="sidebar-link">Examples</a></li><li class="sidebar-sub-header"><a href="/focus-ui/components/button.html#accessibility" class="sidebar-link">Accessibility</a></li></ul></li><li><a href="/focus-ui/components/button-group.html" class="sidebar-link">Button group</a></li><li><a href="/focus-ui/components/card.html" class="sidebar-link">Card</a></li><li><a href="/focus-ui/components/checkbox.html" class="sidebar-link">Checkbox</a></li><li><a href="/focus-ui/components/color-picker.html" class="sidebar-link">Color Picker</a></li><li><a href="/focus-ui/components/data-table.html" class="sidebar-link">Data table</a></li><li><a href="/focus-ui/components/display-text.html" class="sidebar-link">Display text</a></li><li><a href="/focus-ui/components/footer-help.html" class="sidebar-link">Footer help</a></li><li><a href="/focus-ui/components/form.html" class="sidebar-link">Form</a></li><li><a href="/focus-ui/components/form-layout.html" class="sidebar-link">Form layout</a></li><li><a href="/focus-ui/components/frame.html" class="sidebar-link">Frame</a></li><li><a href="/focus-ui/components/heading.html" class="sidebar-link">Heading</a></li><li><a href="/focus-ui/components/icon.html" class="sidebar-link">Icon</a></li><li><a href="/focus-ui/components/layout.html" class="sidebar-link">Layout</a></li><li><a href="/focus-ui/components/link.html" class="sidebar-link">Link</a></li><li><a href="/focus-ui/components/list.html" class="sidebar-link">List</a></li><li><a href="/focus-ui/components/modal.html" class="sidebar-link">Modal</a></li><li><a href="/focus-ui/components/option-list.html" class="sidebar-link">Option list</a></li><li><a href="/focus-ui/components/page.html" class="sidebar-link">Page</a></li><li><a href="/focus-ui/components/pagination.html" class="sidebar-link">Pagination</a></li><li><a href="/focus-ui/components/popover.html" class="sidebar-link">Popover</a></li><li><a href="/focus-ui/components/resource-list.html" class="sidebar-link">Resource list</a></li><li><a href="/focus-ui/components/select.html" class="sidebar-link">Select</a></li><li><a href="/focus-ui/components/skeleton-body-text.html" class="sidebar-link">Skeleton body text</a></li><li><a href="/focus-ui/components/skeleton-display-text.html" class="sidebar-link">Skeleton display text</a></li><li><a href="/focus-ui/components/skeleton-page.html" class="sidebar-link">Skeleton page</a></li><li><a href="/focus-ui/components/skeleton-thumbnail.html" class="sidebar-link">Skeleton thumbnail</a></li><li><a href="/focus-ui/components/spinner.html" class="sidebar-link">Spinner</a></li><li><a href="/focus-ui/components/stack.html" class="sidebar-link">Stack</a></li><li><a href="/focus-ui/components/sub-heading.html" class="sidebar-link">Subheading</a></li><li><a href="/focus-ui/components/text-container.html" class="sidebar-link">Text container</a></li><li><a href="/focus-ui/components/text-field.html" class="sidebar-link">Text Field</a></li><li><a href="/focus-ui/components/text-style.html" class="sidebar-link">Text style</a></li><li><a href="/focus-ui/components/thumbnail.html" class="sidebar-link">Thumbnail</a></li><li><a href="/focus-ui/components/toast.html" class="sidebar-link">Toast</a></li><li><a href="/focus-ui/components/tooltip.html" class="sidebar-link">Tooltip</a></li><li><a href="/focus-ui/components/visually-hidden.html" class="sidebar-link">Visually hidden</a></li></ul></section></li></ul> </aside> <main class="page"> <div class="theme-default-content content__default"><h1 id="button"><a href="#button" aria-hidden="true" class="header-anchor">#</a> Button</h1> <p>Buttons are used to make common ations immediately visible and easy to perform in one click, tap, or keypress.
Users can use them to navigate or to take action.</p> <hr> <h2 id="best-practices"><a href="#best-practices" aria-hidden="true" class="header-anchor">#</a> Best practices</h2> <p>Buttons should:</p> <ul><li>Be clearly and accurately labeled.</li> <li>Lead with strong, actionable verbs.</li> <li>Use established button colors appropriately. For example, only use a red button for an action that's difficult or
impossible to undo.</li> <li>Prioritize the mose important actions. Too many calls to action an cause confusion and make uses unsure of what to
do next.</li> <li>Be positioned in consistent locations on the interface.</li></ul> <hr> <h2 id="examples"><a href="#examples" aria-hidden="true" class="header-anchor">#</a> Examples</h2> <h3 id="basic-button"><a href="#basic-button" aria-hidden="true" class="header-anchor">#</a> Basic button</h3> <div class="language-vue extra-class"><pre class="language-vue"><code><span class="token tag"><span class="token tag"><span class="token punctuation">&lt;</span>Button</span><span class="token punctuation">&gt;</span></span>Add page<span class="token tag"><span class="token tag"><span class="token punctuation">&lt;/</span>Button</span><span class="token punctuation">&gt;</span></span>
</code></pre></div><h3 id="primary-button"><a href="#primary-button" aria-hidden="true" class="header-anchor">#</a> Primary button</h3> <p>Use to highlight the most important actions in any experience. Don't use more than one primary button in a section or
screen to avoid overwhelming users.</p> <div class="language-vue extra-class"><pre class="language-vue"><code><span class="token tag"><span class="token tag"><span class="token punctuation">&lt;</span>Button</span> <span class="token attr-name">primary</span><span class="token punctuation">&gt;</span></span>Save page<span class="token tag"><span class="token tag"><span class="token punctuation">&lt;/</span>Button</span><span class="token punctuation">&gt;</span></span>
</code></pre></div><h3 id="destructive-button"><a href="#destructive-button" aria-hidden="true" class="header-anchor">#</a> Destructive button</h3> <p>Use when the action will delete user data or be otherwise difficult to recover from. Destructive buttons should trigger
a confirmation dialog before the action is completed. Be thoughtful about using destructive buttons because they can
feel stressful for user.</p> <div class="language-vue extra-class"><pre class="language-vue"><code><span class="token tag"><span class="token tag"><span class="token punctuation">&lt;</span>Button</span> <span class="token attr-name">destructive</span><span class="token punctuation">&gt;</span></span>Delete page<span class="token tag"><span class="token tag"><span class="token punctuation">&lt;/</span>Button</span><span class="token punctuation">&gt;</span></span>
</code></pre></div><h3 id="disabled-state"><a href="#disabled-state" aria-hidden="true" class="header-anchor">#</a> Disabled state</h3> <p>Use for actions that aren't currently available. The surrounding interface should make it clear why the button is
disabled and what needs to be done to enable it.</p> <div class="language-vue extra-class"><pre class="language-vue"><code><span class="token tag"><span class="token tag"><span class="token punctuation">&lt;</span>Button</span> <span class="token attr-name">disabled</span><span class="token punctuation">&gt;</span></span>Duplicate page<span class="token tag"><span class="token tag"><span class="token punctuation">&lt;/</span>Button</span><span class="token punctuation">&gt;</span></span>
</code></pre></div><h3 id="loading-state"><a href="#loading-state" aria-hidden="true" class="header-anchor">#</a> Loading state</h3> <p>Use when a button has been pressed and the associated action is in progress.</p> <div class="language-vue extra-class"><pre class="language-vue"><code><span class="token tag"><span class="token tag"><span class="token punctuation">&lt;</span>Button</span> <span class="token attr-name">loading</span><span class="token punctuation">&gt;</span></span>Save page<span class="token tag"><span class="token tag"><span class="token punctuation">&lt;/</span>Button</span><span class="token punctuation">&gt;</span></span>
</code></pre></div><hr> <h2 id="accessibility"><a href="#accessibility" aria-hidden="true" class="header-anchor">#</a> Accessibility</h2> <p>Buttons can have different states that are visually and programmatically conveyed to users.</p> <ul><li>Use the <code>ariaControls</code> prop to add an <code>aria-controls</code> attribute to the button. Use this attribute to point to the
unique <code>id</code> of the content that the button manages.</li> <li>If a button expands or collapses adjacent content, then use the <code>ariaExpanded</code> prop to add the <code>aria-expanded</code>
attribute to the button. Set the value to convey the current expanded (<code>true</code>) or collapsed (<code>false</code>) state of
the content.</li> <li>Use the <code>disabled</code> prop to set the <code>disabled</code> state of the button. This prevents users from being able to interact
with the button, and conveys its inactive state to assistive technologies.</li> <li>Use the <code>ariaPressed</code> prop to add an <code>aria-pressed</code> attribute to the button.</li></ul> <h4 id="navigation"><a href="#navigation" aria-hidden="true" class="header-anchor">#</a> Navigation</h4> <p>Users generally expect buttons to submit data or take action, and for links to navigate. If navigation is required for
the button component, use the <code>url</code> prop. The control will output an anchor styled as a button, instead of a button in
HTML, to help convey this difference.</p> <h3 id="labeling"><a href="#labeling" aria-hidden="true" class="header-anchor">#</a> Labeling</h3> <p>The <code>accessibilityLabel</code> prop adds an <code>aria-label</code> attribute to the button, which can be accessed by assistive
technologies like screen readers. Typically, this label text replaces the visible text on the button for users who
use assistive technologies.</p> <p>Use <code>accessibilityLabel</code> for a button if:</p> <ul><li>The button's visible text doesn't adequately convey the purpose of the button to non-visual users.</li> <li>The button has no text and relies on an icon alone to convey its purpose.</li></ul> <p>To help support users who use speech activation software as well as sighted screen reader users, make sure that the
<code>aria-label</code> text includes any button text that's visible. Mismatches between visible and programmatic labeling can
cause confusion, and might prevent voice recognition commands from working.</p> <p>When possible, give the button visible text that clearly conveys its purpose without the use of <code>accessibilityLabel</code>.
When no additional content is needed, duplicating the button text with <code>accessibilityLabel</code> is not nessesary.</p></div> <footer class="page-edit"><!----> <!----></footer> <div class="page-nav"><p class="inner"><span class="prev">
        ←
        <a href="/focus-ui/components/banner.html" class="prev">
          Banner
        </a></span> <span class="next"><a href="/focus-ui/components/button-group.html">
          Button group
        </a>
        →
      </span></p></div> </main></div><div class="global-ui"></div></div>
    <script src="/focus-ui/assets/js/app.ec7e616f.js" defer></script><script src="/focus-ui/assets/js/2.f856ce09.js" defer></script><script src="/focus-ui/assets/js/10.89c71ea6.js" defer></script>
  </body>
</html>
