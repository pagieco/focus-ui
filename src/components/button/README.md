# Button

Buttons are used to make common ations immediately visible and easy to perform in one click, tap, or keypress.
Users can use them to navigate or to take action.

--- 

## Best practices

Buttons should:

- Be clearly and accurately labeled.
- Lead with strong, actionable verbs.
- Use established button colors appropriately. For example, only use a red button for an action that's difficult or
  impossible to undo.
- Prioritize the mose important actions. Too many calls to action an cause confusion and make uses unsure of what to
  do next.
- Be positioned in consistent locations on the interface.

--- 

## Examples

### Basic button

```vue
<Button>Add page</Button>
```

### Primary button

Use to highlight the most important actions in any experience. Don't use more than one primary button in a section or
screen to avoid overwhelming users.

```vue
<Button primary>Save page</Button>
```

### Destructive button

Use when the action will delete user data or be otherwise difficult to recover from. Destructive buttons should trigger
a confirmation dialog before the action is completed. Be thoughtful about using destructive buttons because they can
feel stressful for user.

```vue
<Button destructive>Delete page</Button>
```

### Disabled state

Use for actions that aren't currently available. The surrounding interface should make it clear why the button is
disabled and what needs to be done to enable it.

```vue
<Button disabled>Duplicate page</Button>
```

### Loading state

Use when a button has been pressed and the associated action is in progress.

```vue
<Button loading>Save page</Button>
```

---

## Accessibility

Buttons can have different states that are visually and programmatically conveyed to users.

- Use the `ariaControls` prop to add an `aria-controls` attribute to the button. Use this attribute to point to the
  unique `id` of the content that the button manages.
- If a button expands or collapses adjacent content, then use the `ariaExpanded` prop to add the `aria-expanded`
  attribute to the button. Set the value to convey the current expanded (`true`) or collapsed (`false`) state of
  the content.
- Use the `disabled` prop to set the `disabled` state of the button. This prevents users from being able to interact
  with the button, and conveys its inactive state to assistive technologies.
- Use the `ariaPressed` prop to add an `aria-pressed` attribute to the button.

#### Navigation

Users generally expect buttons to submit data or take action, and for links to navigate. If navigation is required for 
the button component, use the `url` prop. The control will output an anchor styled as a button, instead of a button in
HTML, to help convey this difference.

### Labeling

The `accessibilityLabel` prop adds an `aria-label` attribute to the button, which can be accessed by assistive
technologies like screen readers. Typically, this label text replaces the visible text on the button for users who
use assistive technologies.

Use `accessibilityLabel` for a button if:

- The button's visible text doesn't adequately convey the purpose of the button to non-visual users.
- The button has no text and relies on an icon alone to convey its purpose.

To help support users who use speech activation software as well as sighted screen reader users, make sure that the
`aria-label` text includes any button text that's visible. Mismatches between visible and programmatic labeling can
cause confusion, and might prevent voice recognition commands from working.

When possible, give the button visible text that clearly conveys its purpose without the use of `accessibilityLabel`.
When no additional content is needed, duplicating the button text with `accessibilityLabel` is not nessesary.
