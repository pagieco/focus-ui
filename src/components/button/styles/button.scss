.button {
  @include button-base;

  color: var(--ink-base);
  background-color: var(--grey-light);
  border: 1px solid var(--grey-dark);

  &:hover {
    background-color: var(--grey-base);
    border-color: var(--grey-dark);
  }
}

.button--primary {
  color: var(--white);
  background-color: var(--blue);
  border-color: var(--blue-dark);

  &:hover {
    background-color: var(--blue-dark);
    border-color: var(--blue-dark);
  }
}

.button--danger {
  color: var(--white);
  background-color: var(--red);
  border-color: var(--red-dark);

  &:hover {
    background-color: var(--red-dark);
    border-color: var(--red-dark);
  }
}

.button--minimal {
  color: var(--blue);
  background-color: transparent;
  border-color: transparent;
  box-shadow: none;

  &:hover {
    color: var(--blue-dark);
    border-color: transparent;
  }
}

.button--small {
  min-height: auto;
  padding: .50em 1em;
  font-size: 80%;
}

.button--disabled {
  pointer-events: none;
  cursor: not-allowed;
  opacity: 0.5;
}

.button--loading {
  .button__label {
    visibility: hidden;
  }
}
