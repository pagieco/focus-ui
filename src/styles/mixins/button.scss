@mixin button-base {
  appearance: none;
  position: relative;
  display: inline-flex;
  align-items: center;
  justify-content: center;
  min-height: 2.9em;
  padding: .75em 2em;
  font-family: inherit;
  font-size: .75rem;
  font-weight: 500;
  line-height: 1;
  text-align: center;
  text-decoration: none;
  white-space: nowrap;
  cursor: pointer;
  border-radius: var(--border-radius);
  outline: none;
  box-shadow: 0 3px 1px 0 rgba(0, 0, 0, 0.075);

  &:active {
    background-color: var(--grey-light);
    box-shadow: inset 0 2px 0 0 rgba(0, 0, 0, 0.075);
  }

  .spinner {
    position: absolute;
  }
}
