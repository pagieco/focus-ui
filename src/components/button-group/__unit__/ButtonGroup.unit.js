import { shallowMount } from '@vue/test-utils';
import { ButtonGroup } from '../index';
import { Button } from '../../button';

const shallowWrap = propsData => shallowMount(ButtonGroup, { propsData });

describe('button-group', () => {
  it('is a valid vue component', () => {
    expect(ButtonGroup).toBeAComponent();
  });

  it('matches the snapshot', () => {
    expect(shallowMount(ButtonGroup).element).toMatchSnapshot();
  });

  it('uses the full-width prop', () => {
    const wrapper = shallowWrap({ fullWidth: true });

    expect(wrapper.classes()).toContain('button-group--fullWidth');
  });

  it('uses the segmented prop', () => {
    const wrapper = shallowWrap({ segmented: true });

    expect(wrapper.classes()).toContain('button-group--segmented');
  });

  it('renders child buttons', () => {
    const wrapper = shallowMount(ButtonGroup, {
      slots: {
        default: Button,
      },
    });

    expect(wrapper.find(Button).isVueInstance()).toBeTruthy();
  });
});
