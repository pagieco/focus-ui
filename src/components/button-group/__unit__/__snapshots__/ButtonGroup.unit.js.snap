// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`button-group matches the snapshot 1`] = `
<div
  class="button-group"
/>
`;
